### PR TITLE
Proposed release workflow

### PR DESCRIPTION
**Pull request by gmaclennan**
## Current release process

> NB: This change recently with https://github.com/digidem/mapeo-mobile/pull/322, this is the process post-#322.

Currently releases are managed by tags: pushing a tag to the repo will cause a release to be built and deployed.

The creation of the git tag should be done with `npm run release` which uses [`https://www.npmjs.com/package/standard-version`](https://www.npmjs.com/package/standard-version) to increment the version in `package.json` (based on parsing conventional commits since the last release to choose the major/minor/patch), create a changelog, and create a tag.

When a tag that matches the pattern `v*.*.*` is pushed to Github it will trigger the [`deploy-release`](https://github.com/digidem/mapeo-mobile/blob/7df584dc6e7972b96d45f6e11d612621dfe7366b/bitrise.yml#L86) workflow, which:

1. Builds release variants of the APK
2. Creates an `.aab` AppBundle for upload to the Play Store
3. Creates a new Github Release and uploads the APK to the release
4. Uploads APKs to S3 for http://mapeo-apks.ddem.us
5. Updates https://mapeo-apks.ddem.us/latest and https://mapeo-apks.ddem.us/icca/latest to point at latest releases

For internal testing before a release, create a tag that matches the pattern `v*.*.*-beta*`, using the script `npm run release:beta`. Beta releases are not uploaded to Github releases and not published on Google Play. They are available as artifacts from the Bitrise build, or on http://mapeo-apks.ddem.us/?prefix=beta-testing/

For developers to test builds, we build a "QA" release for every Pull Request (including PRs from forked repos). These are available on the Bitrise build page https://app.bitrise.io/app/288e6b3c3069b8e6.

## Problems with the current approach

**1. No clear sign-off process**

Creating a release needs to be done by [a developer] on a local machine, and can be done at any time. There are no enforced checks that tests pass, and no way to sign-off on manual QA, e.g. testing offline sync, which is hard/impossible to test automatically.

**2. No clarity about release cycles**

We have discussed the need to have a clear release cycle: a period for adding new features, then a period of internal testing and only fixing bugs, before a release is deemed ready for distribution. All development happens on `master` right now, and the state of the release cycle is held in the team's head.

**3. Confusing patching of critical bugs**

Right now, after a release is published, we begin working on the next release cycle and new features on `master`. If, in the meantime, we find a critical bug in the release that needs to be patched, we need to go back and fork `master` from the last release, and create releases from the fork. This ends up being confusing - releases are made from a variety of different branches.

## Proposed Solution

See this PR's [files](347/files). Please leave comments below or on lines in the file.

### CODE REVIEW

**Review by okdistribute — APPROVED** [review]

File: RELEASES.md
Comment (okdistribute) [inline RELEASES.md]: Shoudl we make 'develop' the 'master' branch then? In GitHub repo settings you can set `develop` as the new master

Comment (gmaclennan) [reply RELEASES.md]: Yes, Github allows you to choose which branch is the default or "base" branch - this is the branch that PRs are automatically made against, unless you change it in the PR (which is what we would do for release PRs, against deploy).

**Comment by gmaclennan** [comment]
## Proposed Implementation

Github provides a lot of tools that enable us to automate and enforce a lot of this process in a way that enables collaboration. I think it is worth switching away from Bitrise for CI and instead using Github Actions as part of implementing this release process for various reasons:

- Github Actions, in addition to running tests, can enforce branch naming, comment on PRs etc.
- Github Actions have no time limits for public repos, and support 40 concurrent tasks (or 5 MacOS concurrent tasks). Bitrise only supports 1 concurrent task, and is limited to 45 minute builds, which we have found times out when running tests.
- Bitrise _requires manual approval for PRs from forked repos_ to manually check that PRs don't try to access secrets. This is a major limitation — contributors need to wait for one of the Dd team to kick off CI each time before they can see if tests pass. Github, instead, blocks secrets from being revealed to PRs from forks, which is another way to achieve the same security.
- Bitrise has better "out-of-the-box" support with different build steps for things like uploading to the Play Store, but there are a [growing number](https://github.com/sdras/awesome-actions) of Github Actions that do the same.
- Bitrise supports hardware acceleration for emulators on Linux, which Github does not, but we can run the Android Emulator tests on the Github MacOS machines, which do support hardware acceleration.

### Protected branches

Github's [protected branches](https://help.github.com/en/github/administering-a-repository/about-protected-branches) are a good way to control access. For the proposed release process:

#### `deploy` branch

This is the strictest - we want to ensure that CI tests pass _and_ a certain number of people have gone through manual QA before we release.

- Require status checks to pass before anything can be merged, and that branches are up-to-date. These status checks can be more than just CI tests, see below.
- Require PR reviews before merging — we can set the number of people who must review before it can merge. This is how we 
- We should include administrators in these rules
- We can optionally restrict pushes to deploy to a specific team - this would be useful if we add more contributors to the repo, but want only a subset of people to be able to trigger a release.

We can also, if it makes sense, require review by [code owners](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) _if_ a PR changes particular code. E.g. we could say someone is the owner of backend mapeo-core code, and require their review if a PR changes something there.

#### `develop` branch

We want all contributions to the `develop` branch to be through PRs, which should pass automated CI tests

- Require status checks to pass
- Do we need code reviews on contributions?

### PR checks

In addition to CI/CD runs, Github Actions allow us to do several other checks on PRs:

#### PRs to `deploy` branch

- Enforce [branch naming](https://github.com/deepakputhraya/action-branch-name) for any PR to `deploy` must be `release/vX.Y` or `release/vX.Y.Z` (this is how we define the version number of the next release).
- Must come from a branch in the `mapeo-mobile` repo (not from a forked repo).
- Ensure the version number in the release PR is semantically correct, by reviewing the `conventional-commit` [recommended bump](https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-recommended-bump)
- Automatically [add a list of reviewers](https://github.com/kentaro-m/auto-assign-action) for any new release PR.
- Reviewers do manual QA, and approve when it passes. The branch protection (see above) enforces a minimum number of people that must do this.
- Ensure that the PR contains a Changelog and a `package.json` version bump to match the version number in the branch name.

#### PRs to `develop` branch

- Can be any name
- Can be from forked repos
- PR title should follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) spec e.g. with https://github.com/aslafy-z/conventional-pr-title-action - this ensures that when squashed and merged we can use the commit message to make the semver bump.
- Optionally: Ensure that the PR is mentioned in a "Release" issue discussion as a way of stopping new PRs being merged to develop while the team is trying to wrap up what should be in the release before cutting the release branch (can then be merged after release branch is cut but before release is deployed).

### CI Builds and tests

It's really useful for CI to be able to generate builds from PRs, so that developers can test new features on a real device before merging, even if we have automated tests set up.

CI like Github Actions and Bitrise offer two types of trigger for a CI run: a `push` to a branch, and a `pull_request`. The `push` event only runs when you create a PR on the same repo - not a PR from a forked repo - and the code that the CI runs on is actually different. `pull_request` CI events run on merged code - the code as it would be if the PR is merged. The `push` event runs on pre-merged code.

#### `pull_request` to `deploy` branch

- Run automated tests (detox / appium) and if possible run robot tests on real devices in the cloud e.g. [Firebase Test Lab](https://firebase.google.com/docs/test-lab).
- Build a beta version and reply to the PR with links to the builds. Use the version from the branch name `release/vX.Y.Z` to name the beta version, along with the build version, so that each commit creates a build with an incrementing number e.g. `v${VERSION}-beta.${BUILD_NUMBER}`.

#### `push` to `deploy` branch

- Add a git tag with version number package.json
- Build a release version and upload to play store and S3 (so APK can be downloaded from website) and Github releases.
- Update release notes with changelog

#### `pull_request` to `develop` branch

We need to be careful with security here. We can't build a version signed with the private key based on a PR, since that would allow someone to open a PR and create a malicious version that could be installed over an existing version and reveal data. We can create a special `QA` version that is signed with a public key.

- Run automated tests (detox / appium) but skip robot tests on real devices (too costly / time consuming for every PR commit, but maybe we can experiment with it)
- Build a QA version and reply to the PR with links to builds.

**Comment by gmaclennan** [comment]
I was also thinking about translations this weekend. CrowdIn has [versions support](https://support.crowdin.com/versions-management/) which can automatically create branches of code and allow translators to work on translations of new features.

I think expecting translators to translate each new feature branch is a bit too much right now, but I think we can integrate translations into the release process, e.g. when we create a new release branch I've configured CrowdIn to automatically sync that branch to CrowdIn, so translators can check any changed translations for the new version.

I'm not sure if we should have some kind of check on this, e.g. do we want to hold off a release until all strings are translated into particular languages? Until we have a larger network this could hold up releases if we are too strict, but also for some users it's not great having some strings still in English. @aliya-ryan  @jencastrodoesstuff ?

**Comment by gmaclennan** [comment]
Looking a bit more into automation options for ensuring that PRs follow conventions. An alternative to using Github Actions for this is to use a Github App. A Github App needs to run on our own infrastructure and is potentially much faster than a Github Action. For the differences see https://jasonet.co/posts/probot-app-or-github-action/.

The advantage of the Github app is that we could write a single app with multiple status checks for our different requirements. There is a nice easy way to write Github Apps with [Probot](https://probot.github.io) and a nice easy way to [deploy them with Glitch](https://github.com/gr2m/glitch-github-app).

I've set up an [example bot](https://glitch.com/~mapeo-pr-probot) on Glitch using [semantic-pull-request](https://github.com/zeke/semantic-pull-requests). This example uses the status API to set the commit status, but it's also possible to use the Checks API which enables more details about why something is failing, which would be useful for contributors. Some examples of using this in a Github App:

- https://github.com/wip/app/blob/master/lib/handle-pull-request-change.js
- https://github.com/fossapps/pr_label_enforcer/blob/master/src/App.ts

An app could check the base branch (`develop` or `deploy`) and then create a status check for each of the checks listed in https://github.com/digidem/mapeo-mobile/pull/347#issuecomment-601823117

**Pull request closed without merging**
